pallets/providers/src/utils.rs

```diff
@@ -830,53 +795,200 @@ where
+            };
+
+            AwaitingTopUpFromProviders::<T>::insert(provider_id, top_up_metadata.clone());
+
+            // Signal to the provider that they need to top up their held deposit to match the current used capacity
+            Self::deposit_event(Event::<T>::AwaitingTopUp {
+                provider_id: *provider_id,
+                top_up_metadata,
+            });
+        }
+
+        // Update the provider's capacity
+        if let Some(mut provider) = MainStorageProviders::<T>::get(provider_id) {
+            provider.capacity = capacity;
+            MainStorageProviders::<T>::mutate(provider_id, |p| *p = Some(provider));
+        } else if let Some(mut provider) = BackupStorageProviders::<T>::get(provider_id) {
+            provider.capacity = capacity;
+            BackupStorageProviders::<T>::mutate(provider_id, |p| *p = Some(provider));
+        }
+
+        Ok(())
+    }
+
+    /// Allows a storage provider to manually top up their held deposit to restore capacity up to their currentlt used capacity.
+    ///
+    /// The provider must be within a grace period due to insufficient capacity.
+    /// Holds the required amount from the provider's free balance to match their used capacity.
+    ///
+    /// This will short circuit if the provider is not registered, doesn't need a top-up, or lacks sufficient balance.
+    pub(crate) fn do_top_up_deposit(account_id: &T::AccountId) -> DispatchResult {
+        let provider_id = AccountIdToMainStorageProviderId::<T>::get(account_id)
+            .or(AccountIdToBackupStorageProviderId::<T>::get(account_id))
+            .ok_or(Error::<T>::NotRegistered)?;
+
+        let (account_id, mut capacity, used_capacity) = Self::get_provider_details(provider_id)?;
+
+        // Need to hold enough balance to increase capacity back to used_capacity
+        let capacity_deficit = used_capacity.saturating_sub(capacity);
+
+        let required_delta_hold_amount =
+            StorageDataUnitAndBalanceConverter::<T>::convert(capacity_deficit);
+
+        // Check if the provider has enough free balance to top up the slashed amount
+        ensure!(
+            T::NativeBalance::can_hold(
+                &HoldReason::StorageProviderDeposit.into(),
+                &account_id,
+                required_delta_hold_amount,
+            ),
+            Error::<T>::CannotHoldDeposit
+        );
+
+        // Hold the slashable amount from the free balance
+        T::NativeBalance::hold(
+            &HoldReason::StorageProviderDeposit.into(),
+            &account_id,
+            required_delta_hold_amount,
+        )?;
+
+        // Set the provider's capacity to the used capacity
+        capacity = used_capacity;
+
+        // Update the provider's capacity in storage
+        if let Some(mut provider) = MainStorageProviders::<T>::get(provider_id) {
+            provider.capacity = capacity;
+            MainStorageProviders::<T>::mutate(provider_id, |p| *p = Some(provider));
```

Comment:
I understand that you're just setting the value, why use `mutate` then?

Reply:
I use `set` now